common/onion.c

```diff
@@ -0,0 +1,305 @@
+	*valid = pull_payload_length(&raw_payload, &max, type, &len);
+
+	if (!*valid)
+		return ROUTING_INFO_SIZE;
+
+	return len;
+}
+
+struct onion_contents *onion_decode(const tal_t *ctx,
+				    const struct route_step *rs)
+{
+	struct onion_contents *c = tal(ctx, struct onion_contents);
+	const u8 *cursor = rs->raw_payload;
+	size_t max = tal_bytelen(cursor), len;
+	struct tlv_tlv_payload *tlv;
+
+	if (!pull_payload_length(&cursor, &max, &c->type, &len))
+		return tal_free(c);
+
+	switch (c->type) {
+	case ONION_V0_PAYLOAD:
+		c->type = ONION_V0_PAYLOAD;
+		c->forward_channel = tal(c, struct short_channel_id);
+		fromwire_short_channel_id(&cursor, &len, c->forward_channel);
+		c->amt_to_forward = fromwire_amount_msat(&cursor, &len);
+		c->outgoing_cltv = fromwire_u32(&cursor, &len);
+		c->payment_secret = NULL;
+
+		if (rs->nextcase == ONION_FORWARD) {
+			c->total_msat = NULL;
+		} else {
+			/* BOLT-e36f7b6517e1173dcbd49da3b516cfe1f48ae556 #4:
+			 * - if it is the final node:
+			 *   - MUST treat `total_msat` as if it were equal to
+			 *     `amt_to_forward` if it is not present. */
+			c->total_msat = tal_dup(c, struct amount_msat,
+						&c->amt_to_forward);
+		}
+		return c;
+
+	case ONION_TLV_PAYLOAD:
+		tlv = tal(c, struct tlv_tlv_payload);
+		if (!fromwire_tlv_payload(&cursor, &len, tlv))
+			return tal_free(c);
+		if (!tlv_payload_is_valid(tlv, NULL))
+			return tal_free(c);
+
+		/* BOLT #4:
+		 *
+		 * The reader:
+		 *   - MUST return an error if `amt_to_forward` or
+		 *     `outgoing_cltv_value` are not present.
+		 */
+		if (!tlv->amt_to_forward || !tlv->outgoing_cltv_value)
+			return tal_free(c);
+
+		amount_msat_from_u64(&c->amt_to_forward,
+				     tlv->amt_to_forward->amt_to_forward);
+		c->outgoing_cltv = tlv->outgoing_cltv_value->outgoing_cltv_value;
+
+		/* BOLT #4:
+		 *
+		 * The writer:
+		 *...
+		 *  - For every non-final node:
+		 *    - MUST include `short_channel_id`
+		 */
+		if (rs->nextcase == ONION_FORWARD) {
+			if (!tlv->short_channel_id)
+				return tal_free(c);
+			c->forward_channel = tal(c, struct short_channel_id);
+			*c->forward_channel
+				= tlv->short_channel_id->short_channel_id;
+			c->total_msat = NULL;
+		} else {
+			c->forward_channel = NULL;
+			/* BOLT-e36f7b6517e1173dcbd49da3b516cfe1f48ae556 #4:
+			 * - if it is the final node:
+			 *   - MUST treat `total_msat` as if it were equal to
+			 *     `amt_to_forward` if it is not present. */
+			c->total_msat = tal_dup(c, struct amount_msat,
+						&c->amt_to_forward);
+		}
+
+		c->payment_secret = NULL;
+
+#if EXPERIMENTAL_FEATURES
+		if (tlv->payment_data) {
+			c->payment_secret = tal_dup(c, struct secret,
+						    &tlv->payment_data->payment_secret);
+			tal_free(c->total_msat);
+			c->total_msat = tal(c, struct amount_msat);
+			c->total_msat->millisatoshis /* Raw: tu64 on wire */
+				= tlv->payment_data->total_msat;
+		}
+#endif
+		tal_free(tlv);
+		return c;
+	}
+
+	/* You said it was a valid type! */
+	abort();
```

Comment:
Did we? I don't see a check anywhere close to the [call on line 975](https://github.com/ElementsProject/lightning/pull/3316/commits/ad9d96821a3c2e920e82e36fd6fbfa999d9290a4#diff-4cff6342b488c21bd84d66ba632d1d4eR975), unless it is covered by a prior call to `pull_payload_length` that I'm missing. Would be good to annotate that on line 975 if that's the case.

Reply:
Yes.  pull_payload_length() is called at the top of this function, and we bail if it returns false.